json/src/read.rs

```diff
@@ -0,0 +1,182 @@
+    iter: LineColIterator<Iter>,
+    /// Temporary storage of peeked byte.
+    ch: Option<u8>,
+}
+
+/// Specialization for Iter=&[u8]. This is more efficient than other iterators
+/// because peek() can be read-only and we can compute line/col position only if
+/// an error happens.
+pub struct SliceRead<'a> {
+    slice: &'a [u8],
+    /// Index of the *next* byte that will be returned by next() or peek().
+    index: usize,
+}
+
+//////////////////////////////////////////////////////////////////////////////
+
+impl<Iter> IteratorRead<Iter>
+    where Iter: Iterator<Item=io::Result<u8>>,
+{
+    pub fn new(iter: Iter) -> Self {
+        IteratorRead {
+            iter: LineColIterator::new(iter),
+            ch: None,
+        }
+    }
+}
+
+impl<Iter> Read for IteratorRead<Iter>
+    where Iter: Iterator<Item=io::Result<u8>>,
+{
+    #[inline]
+    fn next(&mut self) -> io::Result<Option<u8>> {
+        match self.ch.take() {
+            Some(ch) => Ok(Some(ch)),
+            None => {
+                match self.iter.next() {
+                    Some(Err(err)) => Err(err),
+                    Some(Ok(ch)) => Ok(Some(ch)),
+                    None => Ok(None),
+                }
+            }
+        }
+    }
+
+    #[inline]
+    fn peek(&mut self) -> io::Result<Option<u8>> {
+        match self.ch {
+            Some(ch) => Ok(Some(ch)),
+            None => {
+                match self.iter.next() {
+                    Some(Err(err)) => Err(err),
+                    Some(Ok(ch)) => {
+                        self.ch = Some(ch);
+                        Ok(self.ch)
+                    }
+                    None => Ok(None),
+                }
+            }
+        }
+    }
+
+    #[inline]
+    fn discard(&mut self) {
+        self.ch = None;
+    }
+
+    fn position(&self) -> Position {
+        Position {
+            line: self.iter.line(),
+            column: self.iter.col(),
+        }
+    }
+
+    fn peek_position(&self) -> Position {
+        // The LineColIterator updates its position during peek() so it has the
+        // right one here.
+        self.position()
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+
+impl<'a> SliceRead<'a> {
+    pub fn new(slice: &'a [u8]) -> Self {
+        SliceRead {
+            slice: slice,
+            index: 0,
+        }
+    }
+
+    fn position_of_index(&self, i: usize) -> Position {
+        let mut pos = Position { line: 1, column: 0 };
+        for ch in &self.slice[..i] {
+            match *ch {
+                b'\n' => {
+                    pos.line += 1;
+                    pos.column = 0;
+                }
+                _ => {
+                    pos.column += 1;
+                }
+            }
+        }
+        pos
+    }
+}
+
+impl<'a> Read for SliceRead<'a> {
+    #[inline]
+    fn next(&mut self) -> io::Result<Option<u8>> {
+        Ok(if self.index < self.slice.len() {
```

Comment:
Well it is the logic of `self.slice.get(self.index).map(|ch| *ch)`. I tried that and it was consistently about 10% slower. Let me know if it is any different on your computer.

I will add a comment.